sys/include/senml.h

```diff
@@ -0,0 +1,377 @@
+    SENML_UNIT_PARTS_PER_MILLION,           /**< parts per million (ppm, equivalent to 1e-6 '/') */
+    SENML_UNIT_PERCENT,                     /**< percent (/100, equivalent to 1/100 '/') */
+    SENML_UNIT_PERMILLE,                    /**< permille (/1000, equivalent to 1/1000 '/') */
+    SENML_UNIT_HECTOPASCAL,                 /**< hectopascal (hPa, equivalent to 100 Pa) */
+    SENML_UNIT_MILLIMETER,                  /**< millimeter (mm, equivalent to 1/1000 m) */
+    SENML_UNIT_CENTIMETER,                  /**< centimeter (cm, equivalent to 1/100 m) */
+    SENML_UNIT_KILOMETER,                   /**< kilometer (km, equivalent to 1000 m) */
+    SENML_UNIT_KILOMETER_PER_HOUR,          /**< kilometer per hour (km/h, equivalent to 1/3.6 m/s) */
+
+    /* SenML secondary units from CoRE-1 */
+    SENML_UNIT_PARTS_PER_BILLION,   /**< parts per billion (ppb, equivalent to 1e-9 '/') */
+    SENML_UNIT_PARTS_PER_TRILLION,  /**< parts per trillion (ppt, equivalent to 1e-12 '/') */
+    SENML_UNIT_VOLT_AMPERE_HOUR,    /**< volt-ampere-hour (VAh, equivalent to 3600 VAs) */
+    SENML_UNIT_MILLIGRAM_PER_LITER, /**< milligram per liter (mg/l, equivalent to 1/1000 kg/m3) */
+    SENML_UNIT_MICROGRAM_PER_LITER, /**< microgram per liter (ug/l, equivalent to 1e-6 kg/m3) */
+    SENML_UNIT_GRAM_PER_LITER,      /**< gram per liter (g/l, equivalent to 1 kg/m3) */
+} senml_unit_t;
+
+/**
+ * @brief SenML numeric value types.
+ *
+ */
+typedef enum {
+    SENML_TYPE_NUMERIC_UINT,    /**< Unsigned integer */
+    SENML_TYPE_NUMERIC_INT,     /**< Integer */
+    SENML_TYPE_NUMERIC_FLOAT,   /**< Floating point number */
+    SENML_TYPE_NUMERIC_DOUBLE,  /**< Double-precision floating point number */
+    SENML_TYPE_NUMERIC_DECFRAC, /**< Decimal fraction */
+} senml_value_type_t;
+
+/**
+ * @brief Decimal fraction containing a value in the form of m * 10^e.
+ */
+typedef struct {
+    int32_t e;  /**< Exponent */
+    int32_t m;  /**< Mantissa */
+} senml_decfrac_t;
+
+/**
+ * @brief SenML numeric value.
+ *
+ * Various SenML attributes (see @ref senml_attr_t) may contain any 'numeric'
+ * types. This struct is used to contain these.
+ */
+typedef struct {
+    senml_value_type_t type; /**< Type of the value */
+    union {
+        uint64_t u;
+        int64_t i;
+        float f;
+        double d;
+        struct { int32_t e; int32_t m; } df /** Decimal fraction */;
+    } value; /**< Value data */
+} senml_numeric_t;
+
+/**
+ * @brief SenML common record attributes.
+ * All of these values are optional: empty or 0 values will not be encoded.
+ */
+typedef struct {
+    const char *base_name;          /**< Base Name */
+    senml_numeric_t base_time;      /**< Base Time */
+    senml_unit_t base_unit;         /**< Base Unit */
+    senml_numeric_t base_value;     /**< Base Value */
+    senml_numeric_t base_sum;       /**< Base Sum */
+    uint64_t base_version;          /**< Base Version */
```

Comment:
Does it make sense to have compile-time configuration remove some of the optional fields and save some RAM?, specifically `base_version`, `update_time`, `sum`, which will be used more rarely.